File: dotnet/src/SimpleOrm/EntityMapLoader.cs
using System.Collections.Concurrent;

namespace SimpleOrm;

/// <summary>
/// The single entry point for metadata (§7.2). Loader precedence per type:
/// an explicit <see cref="EntityMapBuilder{T}"/> registration, else the attribute
/// loader when any mapping attribute is present, else the convention loader.
/// Maps are cached per loader instance; a type that fails to load throws
/// <see cref="MappingException"/> with every violation.
/// </summary>
public sealed class EntityMapLoader
{
    private readonly MappingOptions _options;
    private readonly ConcurrentDictionary<Type, EntityMap> _cache = new();

    public EntityMapLoader(MappingOptions? options = null) => _options = options ?? MappingOptions.Default;

    public EntityMap Load<T>()
        where T : class
        => Load(typeof(T));

    public EntityMap Load(Type entityType) => _cache.GetOrAdd(entityType, LoadCore);

    private EntityMap LoadCore(Type entityType)
    {
        if (_options.ExplicitMaps.TryGetValue(entityType, out var factory))
        {
            return factory(_options.NamingConvention);
        }

        return AttributeMapLoader.HasMappingAttributes(entityType)
            ? AttributeMapLoader.Load(entityType, _options.NamingConvention)
            : ConventionMapLoader.Load(entityType, _options.NamingConvention);
    }
}

File: php/src/Metadata/EntityMapLoader.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Metadata;

use ReflectionClass;

/**
 * The single entry point for metadata (§7.2). Loader precedence per type: an
 * explicit registration in `MappingOptions::$explicitMaps`, else the attribute
 * loader when any mapping attribute is present, else the convention loader.
 * Maps are cached per instance (CODING-STANDARD §3: no static mutable state —
 * the cache lives here, not in a global). Other areas (session, migrations,
 * validation) call exactly this API.
 */
final class EntityMapLoader
{
    /** @var array<class-string, EntityMap> */
    private array $cache = [];

    private readonly MappingOptions $options;

    public function __construct(?MappingOptions $options = null)
    {
        $this->options = $options ?? MappingOptions::default();
    }

    /** @param class-string $entityType */
    public function load(string $entityType): EntityMap
    {
        return $this->cache[$entityType] ??= $this->loadCore($entityType);
    }

    /** Whether the type carries any mapping attributes (used by tooling such as export-metadata). */
    public static function hasMappingAttributes(string $entityType): bool
    {
        return AttributeMapLoader::hasMappingAttributes(new ReflectionClass($entityType));
    }

    /** @param class-string $entityType */
    private function loadCore(string $entityType): EntityMap
    {
        if (isset($this->options->explicitMaps[$entityType])) {
            return $this->options->explicitMaps[$entityType];
        }

        return self::hasMappingAttributes($entityType)
            ? AttributeMapLoader::load($entityType, $this->options->naming)
            : ConventionMapLoader::load($entityType, $this->options->naming);
    }
}
